src/utils/editor/tests/get-expression.spec.js

```diff
@@ -0,0 +1,183 @@
+
+    it("gets the expression using CodeMirror.getTokenAt", () => {
+      const codemirrorMock = {
+        lineCount: () => 100,
+        getTokenAt: jest.fn(() => ({ start: 0, end: 0 })),
+        doc: {
+          getLine: () => ""
+        }
+      };
+      getExpressionFromCoords(codemirrorMock, { line: 1, column: 1 });
+      expect(codemirrorMock.getTokenAt).toHaveBeenCalled();
+    });
+
+    it("requests the correct line and column from codeMirror", () => {
+      const codemirrorMock = {
+        lineCount: () => 100,
+        getTokenAt: jest.fn(() => ({ start: 0, end: 1 })),
+        doc: {
+          getLine: jest.fn(() => "")
+        }
+      };
+      getExpressionFromCoords(codemirrorMock, { line: 20, column: 5 });
+      // getExpressionsFromCoords uses one based line indexing
+      // CodeMirror uses zero based line indexing
+      expect(codemirrorMock.getTokenAt).toHaveBeenCalledWith({
+        line: 19,
+        ch: 5
+      });
+      expect(codemirrorMock.doc.getLine).toHaveBeenCalledWith(19);
+    });
+
+    it("when called with column 0 returns null", () => {
+      const codemirrorMock = {
+        lineCount: () => 2,
+        getTokenAt: jest.fn(() => ({ start: 0, end: 0, type: null })),
+        doc: {
+          getLine: () => "foo bar;"
+        }
+      };
+
+      const result = getExpressionFromCoords(codemirrorMock, {
+        line: 1,
+        column: 0
+      });
+      expect(codemirrorMock.getTokenAt).toHaveBeenCalledWith({
+        line: 0,
+        ch: 0
+      });
+      expect(result).toBeNull();
+    });
+
+    it("gets the expression when first token on the line", () => {
+      const codemirrorMock = {
+        lineCount: () => 2,
+        getTokenAt: () => ({ start: 0, end: 3 }),
+        doc: {
+          getLine: () => "foo bar;"
+        }
+      };
+
+      const result = getExpressionFromCoords(codemirrorMock, {
+        line: 1,
+        column: 1
+      });
+      expect(result.expression).toEqual("foo");
+      expect(result.location.start).toEqual({ line: 1, column: 0 });
+      expect(result.location.end).toEqual({ line: 1, column: 3 });
+    });
+
+    it("gets the expression when not the first token on the line", () => {
+      const codemirrorMock = {
+        lineCount: () => 2,
+        getTokenAt: () => ({ start: 4, end: 7 }),
+        doc: {
+          getLine: () => "foo bar;"
+        }
+      };
+
+      const result = getExpressionFromCoords(codemirrorMock, {
+        line: 1,
+        column: 5
+      });
+      expect(result.expression).toEqual("bar");
+      expect(result.location.start).toEqual({ line: 1, column: 4 });
+      expect(result.location.end).toEqual({ line: 1, column: 7 });
+    });
+
+    it("includes previous tokens in the expression", () => {
+      const codemirrorMock = {
+        lineCount: () => 2,
+        getTokenAt: location => {
+          if (location.ch >= 5 && location.ch <= 7) {
+            return { start: 4, end: 7, type: "property" };
```

Comment:
I believe in here we are testing `getTokenAt` and not `getExpressionFromCoords`

Reply:
For this I couldn't figure out a nice way to mock the codemirror `getTokenAt` method.  When there are more than one token in an expression this function gets called multiple times and returns different results.  Do you have any suggestions on how I could better mock the `getTokenAt` method?  I had also tried creating a code mirror instance instead of mocking it but I couldn't get it to work in the test environment.